File: javascript/vulnerabilities.js
// Demo các lỗi Vulnerability / Security Hotspot mà SonarQube phát hiện.
// Code DEMO có lỗi cố ý, KHÔNG dùng cho production.

const https = require("https");
const { exec } = require("child_process");
const crypto = require("crypto");

// Security Hotspot / Vulnerability: hardcoded credentials
const DB_PASSWORD = "SuperSecret123!";
const API_KEY = "FAKE_DEMO_API_KEY_NOT_REAL";

// Vulnerability: tắt xác thực chứng chỉ TLS toàn cục
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

function getUser(db, username) {
  // SQL Injection: nối trực tiếp input vào query
  const query = "SELECT * FROM users WHERE name = '" + username + "'";
  return db.query(query);
}

function pingHost(host) {
  // Command Injection: exec với input người dùng
  exec("ping -c 1 " + host, (err, stdout) => {
    console.log(stdout);
  });
}

function runCode(userInput) {
  // Code Injection: eval trên dữ liệu đầu vào
  return eval(userInput);
}

function generateToken() {
  // Security Hotspot: Math.random() không an toàn cho token bảo mật
  return Math.random().toString(36).substring(2);
}

function weakHash(password) {
  // Weak crypto: MD5 không an toàn
  return crypto.createHash("md5").update(password).digest("hex");
}

function renderProfile(userInput) {
  // DOM XSS: gán trực tiếp input vào innerHTML
  const el = document.getElementById("profile");
  el.innerHTML = userInput;
}

function fetchInsecure(url) {
  // Insecure: tắt xác thực chứng chỉ
  const agent = new https.Agent({ rejectUnauthorized: false });
  return fetch(url, { agent });
}

// ===== Lỗi MỚI chèn vào code cũ (key giả, không phải secret thật) =====
const PAYPAL_SECRET = "FAKE_paypal_live_secret_NOTREAL";
const ENCRYPTION_KEY = "FAKE_aes_key_1234567890_NOTREAL";

function buildQuery(table, userId) {
  // SQL Injection MOI
  return "SELECT * FROM " + table + " WHERE id = " + userId;
}

function insecureRandomPassword() {
  // Security Hotspot MOI: Math.random cho mat khau
  return "pwd_" + Math.random().toString(36).slice(2);
}

function openRedirect(res, url) {
  // Vulnerability MOI: redirect theo input khong kiem tra
  res.writeHead(302, { Location: url });
  res.end();
}

module.exports = {
  getUser,
  pingHost,
  runCode,
  generateToken,
  weakHash,
  renderProfile,
  fetchInsecure,
  buildQuery,
  insecureRandomPassword,
  openRedirect,
  DB_PASSWORD,
  API_KEY,
  PAYPAL_SECRET,
  ENCRYPTION_KEY,
};

File: python/vulnerabilities.py
"""Demo các lỗi Vulnerability / Security Hotspot mà SonarQube phát hiện.

Đây là code DEMO có lỗi cố ý, KHÔNG dùng cho production.
"""

import hashlib
import os
import sqlite3
import subprocess

import requests

# Security Hotspot / Vulnerability: hardcoded credentials
DB_PASSWORD = "SuperSecret123!"
API_KEY = "FAKE_DEMO_API_KEY_NOT_REAL"
AWS_SECRET = "FAKE_DEMO_AWS_SECRET_NOT_REAL"


def get_user(username):
    """SQL Injection: nối trực tiếp input vào câu query."""
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE name = '" + username + "'"
    cursor.execute(query)
    return cursor.fetchall()


def ping_host(host):
    """Command Injection: shell=True với input người dùng."""
    cmd = "ping -c 1 " + host
    return subprocess.call(cmd, shell=True)


def run_code(user_input):
    """Code Injection: eval trên dữ liệu đầu vào."""
    return eval(user_input)


def hash_password(password):
    """Weak crypto: MD5 không an toàn cho mật khẩu."""
    return hashlib.md5(password.encode()).hexdigest()


def fetch_data(url):
    """Insecure: tắt xác thực chứng chỉ SSL/TLS."""
    response = requests.get(url, verify=False)
    return response.text


def connect_db():
    """Hardcoded credentials sử dụng trực tiếp."""
    return f"postgres://admin:{DB_PASSWORD}@10.0.0.5:5432/prod"


def read_secret_from_env():
    """So sánh: cách làm đúng là đọc từ biến môi trường (để đối chiếu)."""
    return os.environ.get("API_KEY", "")
